messages.py

```diff
@@ -0,0 +1,278 @@
+    def __init__(self, value):
+        self.value = value
+
+    def asdict(self):
+        return {
+            'type': 'STATE',
+            'value': self.value
+        }
+
+
+class ActivateVersionMessage(Message):
+    '''ACTIVATE_VERSION message (EXPERIMENTAL).
+
+    The ACTIVATE_VERSION messages has these fields:
+
+      * stream - The name of the stream.
+      * version - The version number to activate.
+
+    This is a signal to the Target that it should delete all previously
+    seen data and replace it with all the RECORDs it has seen where the
+    record's version matches this version number.
+
+    Note that this feature is experimental. Most Taps and Targets should
+    not need to use the "version" field of "RECORD" messages or the
+    "ACTIVATE_VERSION" message at all.
+
+    >>> msg = singer.ActivateVersionMessage(
+    >>>     stream='users',
+    >>>     version=2)
+
+    '''
+    def __init__(self, stream, version):
+        self.stream = stream
+        self.version = version
+
+    def asdict(self):
+        return {
+            'type': 'ACTIVATE_VERSION',
+            'stream': self.stream,
+            'version': self.version
+        }
+
+
+def _required_key(msg, k):
+    if k not in msg:
+        raise Exception("Message is missing required key '{}': {}".format(k, msg))
+
+    return msg[k]
+
+
+def parse_message(msg):
+    """Parse a message string into a Message object."""
+    obj = json.loads(msg)
+    msg_type = _required_key(obj, 'type')
+
+    if msg_type == 'RECORD':
+        time_extracted = obj.get('time_extracted')
+        repo_path = obj.get('repo_path')
+        if time_extracted:
+            time_extracted = dateutil.parser.parse(time_extracted)
+        return RecordMessage(stream=_required_key(obj, 'stream'),
+                             record=_required_key(obj, 'record'),
+                             version=obj.get('version'),
+                             time_extracted=time_extracted,
+                             repo_path=repo_path)
+
+
+    elif msg_type == 'SCHEMA':
+        return SchemaMessage(stream=_required_key(obj, 'stream'),
+                             schema=_required_key(obj, 'schema'),
+                             key_properties=_required_key(obj, 'key_properties'),
+                             bookmark_properties=obj.get('bookmark_properties'))
+
+    elif msg_type == 'STATE':
+        return StateMessage(value=_required_key(obj, 'value'))
+
+    elif msg_type == 'ACTIVATE_VERSION':
+        return ActivateVersionMessage(stream=_required_key(obj, 'stream'),
+                                      version=_required_key(obj, 'version'))
+    else:
+        return None
+
+
+def format_message(message):
+    return json.dumps(message.asdict(), use_decimal=True)
+
+
+def write_message(message):
+    sys.stdout.write(format_message(message) + '\n')
+    sys.stdout.flush()
+
+
+def write_record(stream_name, record, stream_alias=None, time_extracted=None, repo_path=None):
```

Comment:
so I considered that but I wasn't sure if it was better to make it part of the "meta" data (like extracted_time). Actually writing it to the message chunk is easier. @briansloane do you want me to take  a crack at that?

Reply:
Ah right. I'll let @briansloane weigh in on how you all should divide up the work but FWIW metadata isn't persisted anywhere generally. So in this case the hope was to make it so that users of the resulting data could refer to the source repo in queries etc. so metadata is unambiguously the wrong place to store it. NBD either way. :)